### PR TITLE
Refactor: make SSrunechat to be subsystem of SStimer

### PR DESCRIPTION
**Pull request by Bizzonium**
## About The Pull Request

Refactoring to get rid of complex buggy duplicate code. At some point i gave up on porting fixes to runechat so i just made runechat a subsystem of SStimer.

## Why It's Good For The Game

Less duplicate code. Easier to maintain.

## Changelog
:cl:
refactor: make SSrunechat to be subsystem of SStimer 
/:cl:

### CODE REVIEW

**Comment by bobbahbrown** [comment]
have you tested how this affects performance on large-population servers? the reason i made it a separate subsystem was to mimic what the timer subsystem does while stripping out what unnecessary overhead may exist. namely creation and tracking of unnecessary timers, as this has caused the timer subsystem to become overwhelmed and fail to recover in the past.

best wishes,
bobbah 'bee' brown (creator of runechat)

**Comment by bobbahbrown** [comment]
for sake of being able to understand the change prior, see https://github.com/tgstation/tgstation/pull/52425

**Comment by bobbahbrown** [comment]
unless im clinically insane (possible) that change was to fix somewhat prevalent issues with the timer subsystem collapsing on itself and essentially stopping when it filled up with an absurd amount of timers from a ton of people talking.

**Comment by Bizzonium** [comment]
@bobbahbrown hello! Since we fixed themajor SStimer issues in the last six months it became stable as never before: 
https://github.com/tgstation/tgstation/pull/59718
https://github.com/tgstation/tgstation/pull/63427
https://github.com/tgstation/tgstation/pull/64138
https://github.com/tgstation/tgstation/pull/64242

https://github.com/tgstation/tgstation/pull/60846 - This PR made possible to increase chances of MC recovery 
https://github.com/tgstation/tgstation/pull/59073 - This PR allowed us to create subsystems based on SStimer without code duplication 

Here is results for TGMC with 100+ pop. Before avg time dilation was around 10% and 40+% if runechat broke. Now it's around 4%.

image
image

**Comment by Bizzonium** [comment]
Yes, test merge would be great. Maybe another issues will come out

File: code/datums/chatmessage.dm [hunk 1]
```diff
@@ -208,20 +214,20 @@
...
  *
  * Arguments:
  * * fadetime - The amount of time to animate the message's fadeout for
  */
 /datum/chatmessage/proc/end_of_life(fadetime = CHAT_MESSAGE_EOL_FADE)
-	eol_complete = scheduled_destruction + fadetime
+	isFading = TRUE
 	animate(message, alpha = 0, time = fadetime, flags = ANIMATION_PARALLEL)
-	enter_subsystem(eol_complete) // re-enter the runechat SS with the EOL completion time to QDEL self
+	addtimer(CALLBACK(src, .proc/Destroy), fadetime, TIMER_DELETE_ME, SSrunechat)
```
Comment (Bizzonium) [inline code/datums/chatmessage.dm:230]: Should we use QDEL_IN here? Or this solution is ok?

File: code/datums/chatmessage.dm [hunk 2]
```diff
@@ -180,11 +182,15 @@
...
-			// scheduled time once the EOL completion time has been set.
-			var/sched_remaining = m.scheduled_destruction - world.time
-			if (!m.eol_complete)
+			// scheduled time once the EOL has been executed.
+			if (!m.isFading)
+				var/sched_remaining = timeleft(m.fadertimer)
 				var/remaining_time = (sched_remaining) * (CHAT_MESSAGE_EXP_DECAY ** idx++) * (CHAT_MESSAGE_HEIGHT_DECAY ** combined_height)
-				m.enter_subsystem(world.time + remaining_time) // push updated time to runechat SS
+				if (remaining_time)
+					deltimer(m.fadertimer, SSrunechat)
```
Comment (Bizzonium) [inline code/datums/chatmessage.dm:190]: Should we use TIMER_UNIQUE|TIMER_OVERRIDE flags for timer and just override it?

File: code/datums/chatmessage.dm (hunk 1 above)
Comment (Mothblocks) [reply code/datums/chatmessage.dm:230]: Not using QDEL_IN is fine if you want it to be on SSrunechat (though we could make a new macro for that), butdon't call `Destroy` directly at the very least. Use `GLOBAL_PROC, /proc/qdel, src` instead.

Comment (Bizzonium) [reply code/datums/chatmessage.dm]: https://github.com/tgstation/tgstation/pull/64366/commits/bd6a9a1645c88b874434e78eec37e246e3673f1c

**Comment by Mothblocks** [comment]
TMing on Manuel

**Comment by Mothblocks** [comment]
I haven't heard any bugs

**Pull request merged**